common/src/main/java/io/netty/util/ResourceLeakDetector.java

```diff
@@ -333,18 +334,29 @@ protected void reportInstancesLeak(String resourceType) {
     }
 
     @SuppressWarnings("deprecation")
-    private final class DefaultResourceLeak extends PhantomReference<Object> implements ResourceLeakTracker<T>,
-            ResourceLeak {
-        private final Record head;
+    private static final class DefaultResourceLeak<T>
+            extends PhantomReference<Object> implements ResourceLeakTracker<T>, ResourceLeak {
 
-        // This will be updated once a new Record will be created and added
-        private Record tail;
+        @SuppressWarnings("unchecked") // generics and updaters do not mix.
+        private static final AtomicReferenceFieldUpdater<DefaultResourceLeak<?>, Record> headUpdater =
```

Comment:
all the joy of java generics

Reply:
I tried for at least 30 minutes to get ARFU to match the generics on the inner class.   I gave up and went with the cast.  The problem is mainly that DefaultResourceLeak.class is static and has the wrong type.

Reply:
no worries :)